File: sdk/go/temporalstore/examples/proxy/sequence_features.go
package main

import (
	"fmt"
	"log"

	temporalstore "temporalstore"
)

func main() {
	client := temporalstore.ConnectProxy(temporalstore.ProxyOptions{
		Endpoint:      "http://127.0.0.1:8080",
		NamespaceName: "sdk_ns",
		TableName:     "sdk_table",
	})

	if err := client.PutString("go:proxy:user:42", `{"tier":"gold"}`, 0); err != nil {
		log.Fatal(err)
	}
	profile, err := client.GetString("go:proxy:user:42")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("profile=%s\n", profile)

	key := "go:proxy:user:42:sequence"
	err = client.AddSequenceFeatureRows(key, []temporalstore.SequenceFeatureRow{
		{Timestamp: 1700000000000, GID: 900, ActionType: 1, Duration: 31, AuthorID: 7000},
		{Timestamp: 1700000001000, GID: 901, ActionType: 3, Duration: 120, AuthorID: 7001},
	})
	if err != nil {
		log.Fatal(err)
	}

	rows, err := client.QuerySequenceFeatureRows(
		key,
		1700000000000,
		1700000002000,
		10,
		[]temporalstore.FeatureFilter{
			{Field: "action_type", Op: temporalstore.FeatureFilterEqual, Value: 3},
		},
	)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("rows=%+v\n", rows)
}

File: sdk/java/temporalstore/src/main/java/com/temporalstore/example/SequenceFeatures.java
package com.temporalstore.example;

import com.temporalstore.TemporalStoreClient;
import java.util.Arrays;
import java.util.List;

public final class SequenceFeatures {
    private SequenceFeatures() {}

    public static void main(String[] args) {
        TemporalStoreClient.Options options = new TemporalStoreClient.Options();
        options.libraryName = System.getenv().getOrDefault("TEMPORALSTORE_JAVA_LIB", "bcache2");
        options.metaserverAddr = args.length > 0 ? args[0] : "127.0.0.1:18200";
        options.namespaceName = args.length > 1 ? args[1] : "sdk_ns";
        options.tableName = args.length > 2 ? args[2] : "sdk_table";
        options.psm = "temporalstore.java.example";

        try (TemporalStoreClient client = TemporalStoreClient.connect(options)) {
            client.putString("java:user:42", "{\"tier\":\"gold\"}");
            System.out.println("profile=" + client.getString("java:user:42"));
            client.expire("java:user:42", 60000);
            System.out.println("ttl_ms=" + client.ttl("java:user:42"));

            client.hset("java:user:42:features", "ctr_7d", "0.042");
            System.out.println("ctr_7d=" + client.hget("java:user:42:features", "ctr_7d"));

            client.sadd("java:user:42:campaigns", "campaign_100");
            System.out.println("campaigns=" + client.smembers("java:user:42:campaigns"));

            String key = "java:user:42:sequence";
            client.addSequenceFeatureRows(
                    key,
                    Arrays.asList(
                            new TemporalStoreClient.SequenceFeatureRow(
                                    1700000000000L, 900L, 1, 31, 7000L),
                            new TemporalStoreClient.SequenceFeatureRow(
                                    1700000001000L, 901L, 3, 120, 7001L)));

            List<TemporalStoreClient.SequenceFeatureRow> rows =
                    client.querySequenceFeatureRows(
                            key,
                            1700000000000L,
                            1700000002000L,
                            10,
                            Arrays.asList(
                                    new TemporalStoreClient.FeatureFilter(
                                            "action_type",
                                            TemporalStoreClient.FeatureFilterOp.EQUAL,
                                            3)));
            System.out.println("rows=" + rows.size());

            String riskKey = "java:user:42:risk";
            client.riskIncrement(
                    riskKey,
                    1,
                    24 * 3600,
                    TemporalStoreClient.RiskPrecision.ONE_MINUTE,
                    "java-risk-1",
                    0);
            System.out.println(
                    "risk_count="
                            + client.riskCount(
                                    riskKey,
                                    TemporalStoreClient.RiskPrecision.ONE_MINUTE,
                                    -1,
                                    0,
                                    TemporalStoreClient.WindowUnit.HOUR));
        }
    }
}
